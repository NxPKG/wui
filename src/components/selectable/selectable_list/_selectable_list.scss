// Expand height of list via flex box
.wuiSelectableList {
  &:focus-within {
    @include wuiFocusRing;
  }
}

.wuiSelectableList-fullHeight {
  flex-grow: 1;
}

.wuiSelectableList-bordered {
  overflow: hidden;
  border: $wuiBorderThin;
  border-radius: $wuiBorderRadius;
}

.wuiSelectableList__list {
  @include wuiOverflowShadow;
  @include wuiScrollBar;
}

.wuiSelectableList__groupLabel {
  @include wuiTitle('xxxs');
  display: flex;
  align-items: center;
  border-bottom: $wuiSelectableListItemBorder;
  padding: $wuiSelectableListItemPadding;
}
