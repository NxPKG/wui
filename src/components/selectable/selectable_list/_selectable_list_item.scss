.wuiSelectableListItem {
  @include wuiFontSizeS;
  display: inline-flex; // Necessary to make sure it doesn't force the whole popover to be too wide
  width: 100%;
  text-align: left;
  color: $wuiTextColor;
  cursor: pointer;

  &:not(:last-of-type) {
    border-bottom: $wuiSelectableListItemBorder;
  }

  &-isFocused:not([aria-disabled='true']),
  &:hover:not([aria-disabled='true']) {
    color: $wuiColorPrimary;
    background-color: $wuiFocusBackgroundColor;

    .wuiSelectableListItem__text {
      text-decoration: underline;
    }
  }

  &[aria-disabled='true'] {
    color: $wuiColorMediumShade;
    cursor: not-allowed;
  }
}

.wuiSelectableListItem__content {
  padding: $wuiSelectableListItemPadding;
  width: 100%;
  display: flex;
  align-items: center;
}

.wuiSelectableListItem__icon,
.wuiSelectableListItem__prepend {
  margin-right: $wuiSizeM;
  flex-shrink: 0;
}

.wuiSelectableListItem__append {
  margin-left: $wuiSizeM;
  flex-shrink: 0;
}

.wuiSelectableListItem__text {
  @include wuiTextTruncate;
  flex-grow: 1; // Pushes appended content to the far right
}
