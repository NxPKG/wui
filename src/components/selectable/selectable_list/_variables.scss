$wuiSelectableListItemBorder: 1px solid darken($wuiColorLightestShade, 2%);
$wuiSelectableListItemPadding: $wuiSizeXS $wuiSizeM;
