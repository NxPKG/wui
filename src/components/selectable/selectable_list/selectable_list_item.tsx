/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import classNames from 'classnames';
import React, { Component, LiHTMLAttributes } from 'react';
import { CommonProps } from '../../common';
import { WuiI18n } from '../../i18n';
import { WuiIcon, IconColor, IconType } from '../../icon';
import { WuiSelectableOptionCheckedType } from '../selectable_option';
import { WuiScreenReaderOnly } from '../../accessibility';
import { WuiBadge, WuiBadgeProps } from '../../badge';

function resolveIconAndColor(
  checked: WuiSelectableOptionCheckedType
): { icon: IconType; color?: IconColor } {
  if (!checked) {
    return { icon: 'empty' };
  }
  return checked === 'on'
    ? { icon: 'check', color: 'text' }
    : { icon: 'cross', color: 'text' };
}

export type WuiSelectableListItemProps = LiHTMLAttributes<HTMLLIElement> &
  CommonProps & {
    children?: React.ReactNode;
    /**
     * Applies an icon and visual styling to activated items
     */
    checked?: WuiSelectableOptionCheckedType;
    /**
     * Shows icons based on `checked` type
     */
    showIcons: boolean;
    /**
     * Highlights the item for pseudo focus
     */
    isFocused?: boolean;
    disabled?: boolean;
    prepend?: React.ReactNode;
    append?: React.ReactNode;
    allowExclusions?: boolean;
    /**
     * When enabled by setting to either `true` or passing custom a custom badge,
     * shows a hollow badge as an append (far right) when the item is focused.
     * The default content when `true` is `â†© to select/deselect/include/exclude`
     */
    onFocusBadge?: boolean | WuiBadgeProps;
  };

// eslint-disable-next-line react/prefer-stateless-function
export class WuiSelectableListItem extends Component<
  WuiSelectableListItemProps
> {
  static defaultProps = {
    showIcons: true,
    onFocusBadge: true,
  };

  constructor(props: WuiSelectableListItemProps) {
    super(props);
  }

  render() {
    const {
      children,
      className,
      disabled,
      checked,
      isFocused,
      showIcons,
      prepend,
      append,
      allowExclusions,
      onFocusBadge,
      ...rest
    } = this.props;

    const classes = classNames(
      'wuiSelectableListItem',
      {
        'wuiSelectableListItem-isFocused': isFocused,
      },
      className
    );

    let optionIcon: React.ReactNode;
    if (showIcons) {
      const { icon, color } = resolveIconAndColor(checked);
      optionIcon = (
        <WuiIcon
          className="wuiSelectableListItem__icon"
          color={color}
          type={icon}
        />
      );
    }

    let state: React.ReactNode;
    let instruction: React.ReactNode;
    if (allowExclusions && checked === 'on') {
      state = (
        <WuiScreenReaderOnly>
          <span>
            <WuiI18n
              token="wuiSelectableListItem.includedOption"
              default="Included option."
            />
          </span>
        </WuiScreenReaderOnly>
      );
      instruction = (
        <WuiScreenReaderOnly>
          <span>
            <WuiI18n
              token="wuiSelectableListItem.includedOptionInstructions"
              default="To exclude this option, press enter."
            />
          </span>
        </WuiScreenReaderOnly>
      );
    } else if (allowExclusions && checked === 'off') {
      state = (
        <WuiScreenReaderOnly>
          <span>
            <WuiI18n
              token="wuiSelectableListItem.excludedOption"
              default="Excluded option."
            />
          </span>
        </WuiScreenReaderOnly>
      );
      instruction = (
        <WuiScreenReaderOnly>
          <span>
            <WuiI18n
              token="wuiSelectableListItem.excludedOptionInstructions"
              default="To deselect this option, press enter."
            />
          </span>
        </WuiScreenReaderOnly>
      );
    }

    let prependNode: React.ReactNode;
    if (prepend) {
      prependNode = (
        <span className="wuiSelectableListItem__prepend">{prepend}</span>
      );
    }

    let appendNode: React.ReactNode;
    if (append || !!onFocusBadge) {
      let onFocusBadgeNode: React.ReactNode;
      const defaultOnFocusBadgeProps: WuiBadgeProps = {
        'aria-hidden': true,
        iconType: 'returnKey',
        iconSide: 'left',
        color: 'hollow',
      };

      if (onFocusBadge === true) {
        onFocusBadgeNode = (
          <WuiBadge
            className="wuiSelectableListItem__onFocusBadge"
            {...defaultOnFocusBadgeProps}
          />
        );
      } else if (!!onFocusBadge && onFocusBadge !== false) {
        const { children, className, ...restBadgeProps } = onFocusBadge;
        onFocusBadgeNode = (
          <WuiBadge
            className={classNames(
              'wuiSelectableListItem__onFocusBadge',
              className
            )}
            {...defaultOnFocusBadgeProps}
            {...(restBadgeProps as WuiBadgeProps)}>
            {children}
          </WuiBadge>
        );
      }

      // Only display the append wrapper if append exists or isFocused
      if (append || (isFocused && !disabled)) {
        appendNode = (
          <span className="wuiSelectableListItem__append">
            {append} {isFocused && !disabled ? onFocusBadgeNode : null}
          </span>
        );
      }
    }

    return (
      <li
        // eslint-disable-next-line jsx-a11y/no-noninteractive-element-to-interactive-role
        role="option"
        aria-selected={!disabled && typeof checked === 'string'}
        className={classes}
        aria-disabled={disabled}
        {...rest}>
        <span className="wuiSelectableListItem__content">
          {optionIcon}
          {prependNode}
          <span className="wuiSelectableListItem__text">
            {state}
            {children}
            {instruction}
          </span>
          {appendNode}
        </span>
      </li>
    );
  }
}
