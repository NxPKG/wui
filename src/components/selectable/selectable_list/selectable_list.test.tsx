/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../../test/required_props';

import { WuiSelectableList } from './selectable_list';
import { WuiSelectableOption } from '../selectable_option';

// Mock the htmlIdGenerator to generate predictable ids for snapshot tests
jest.mock('../../../services/accessibility/html_id_generator', () => ({
  htmlIdGenerator: () => () => 'htmlId',
}));

const options: WuiSelectableOption[] = [
  {
    label: 'Titan',
    'data-test-subj': 'titanOption',
  },
  {
    label: 'Enceladus',
  },
  {
    label: 'Mimas',
  },
  {
    label:
      "Pandora is one of Saturn's moons, named for a Titaness of Greek mythology",
  },
  {
    label: 'Tethys',
  },
  {
    label: 'Hyperion',
  },
];

const selectableListRequiredProps = {
  makeOptionId: (index: number | undefined) => `option_${index}`,
  listId: 'list',
  onOptionClick: () => {},
  setActiveOptionIndex: () => {},
  ...requiredProps,
};

describe('WuiSelectableListItem', () => {
  test('is rendered', () => {
    const component = render(
      <WuiSelectableList options={options} {...selectableListRequiredProps} />
    );

    expect(component).toMatchSnapshot();
  });
  describe('props', () => {
    test('visibleOptions', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          visibleOptions={options.slice(2)}
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('searchValue', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          searchValue="Mi"
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('searchValue', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          searchValue="Mi"
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('renderOption', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          renderOption={(option: WuiSelectableOption, searchValue?: string) => {
            return (
              <span>
                {searchValue} =&gt; {option.label}
              </span>
            );
          }}
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('height is forced', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          height={200}
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('height is full', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          height="full"
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('allowExclusions', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          allowExclusions
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('activeOptionIndex', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          activeOptionIndex={2}
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('rowHeight', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          rowHeight={20}
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('showIcons can be turned off', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          showIcons={false}
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('singleSelection can be turned on', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          singleSelection={true}
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('singleSelection can be forced so that at least one must be selected', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          singleSelection="always"
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });

    test('bordered', () => {
      const component = render(
        <WuiSelectableList
          options={options}
          bordered
          {...selectableListRequiredProps}
        />
      );

      expect(component).toMatchSnapshot();
    });
  });
});
