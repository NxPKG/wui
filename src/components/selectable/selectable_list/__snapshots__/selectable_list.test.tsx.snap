// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WuiSelectableListItem is rendered 1`] = `
<div
  class="wuiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props activeOptionIndex 1`] = `
<div
  class="wuiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props allowExclusions 1`] = `
<div
  class="wuiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props bordered 1`] = `
<div
  class="wuiSelectableList wuiSelectableList-bordered testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props height is forced 1`] = `
<div
  class="wuiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props height is full 1`] = `
<div
  class="wuiSelectableList wuiSelectableList-fullHeight testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;height:0;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props renderOption 1`] = `
<div
  class="wuiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props rowHeight 1`] = `
<div
  class="wuiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props searchValue 1`] = `
<div
  class="wuiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props searchValue 2`] = `
<div
  class="wuiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props showIcons can be turned off 1`] = `
<div
  class="wuiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props singleSelection can be forced so that at least one must be selected 1`] = `
<div
  class="wuiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props singleSelection can be turned on 1`] = `
<div
  class="wuiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;

exports[`WuiSelectableListItem props visibleOptions 1`] = `
<div
  class="wuiSelectableList testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    style="overflow:visible;width:0"
  />
</div>
`;
